platform/solr/solr-factory/src/test/groovy/org/codice/solr/factory/impl/SolrClientAdapterAsyncSpec.groovy

```diff
@@ -0,0 +1,1253 @@
+      adapter.available
+
+    when: "calling any api methods"
+      adapter.request(request, COLLECTION)
+
+    then: "have it suddenly fail"
+      1 * client.request(request, COLLECTION) >> {
+        throw exception.fillInStackTrace()
+      }
+      def e = thrown(Exception)
+
+      e.is(exception)
+
+    and: "the adapter should be unavailable because it failed to connect to the client after the error"
+      !adapter.available
+
+    when: "calling any api methods after that should fail right away"
+      adapter.request(request, COLLECTION)
+
+    then: "verify it didn't get delegated to the underlying client"
+      0 * client.request(request, COLLECTION)
+
+    and: "it threw an unavailable exception indicating we are unavailable because of that failed first api call"
+      e = thrown(UnavailableSolrException)
+      e.cause.is(exception)
+
+    when: "notifying the ping controller to connect"
+      pingController.notifyTo('connect')
+
+    and: "waiting for failsafe to complete"
+      pingController.waitForSuccessfulCompletion()
+
+    then: "the adapter should be available again"
+      adapter.available
+
+    when: "calling any api methods"
+      def returnedResponse = adapter.request(request, COLLECTION)
+
+    then: "verify the underlying client was called"
+      1 * client.request(request, COLLECTION) >> response
+      returnedResponse.is(response)
+
+    and: "verify failsafe controllers"
+      createController.verify()
+      pingController.verify()
+
+    cleanup:
+      createController?.shutdown()
+      pingController?.shutdown()
+
+    where:
+      exception << [new SolrException(ErrorCode.UNKNOWN, 'testing request'), new SolrServerException('testing'), new IOException('testing')]
+  }
+
+  def 'test API calls failing with RuntimeException will not move the state to unavailable and will not try to reconnect'() {
+    given: "a client"
+      def client = Mock(SolrClient)
+
+    and: "controllers that mocks every execution/attempts"
+      def createController = new FailsafeController('SolrClient Creation').onNextExecution(doReturn(client))
+      def pingController = new FailsafeController('SolrClient Ping').onNextExecution(doReturn())
+
+    and:
+      def exception = new RuntimeException()
+
+    when: "creating an adapter"
+      def adapter = new SolrClientAdapter(CORE, creator, createController.&with, pingController.&with)
+
+    and: "waiting for failsafe to complete"
+      createController.waitForSuccessfulCompletion()
+      pingController.waitForSuccessfulCompletion()
+
+    then: "the adapter should be available"
+      adapter.available
+
+    when: "calling any api methods"
+      adapter.request(request, COLLECTION)
+
+    then: "have it suddenly fail"
+      1 * client.request(request, COLLECTION) >> {
+        throw exception.fillInStackTrace()
+      }
+      def e = thrown(Exception)
+
+      e.is(exception)
+
+    and: "cthe adapter should still be available"
+      adapter.available
```

Comment:
✏️ Should we be consistent between tests? In the previous one, we checked both `isAvailable` (L406) and that calls were still delegated to the `client` (L412), but here we only check `isAvailable`. Having separate tests that verify the state and others that verify the call delegation would probably help here. The same way we're not testing the listener and initializer notifications at the same time as we test everything else.

Reply:
The first one is verifying that the adapter was first available before we proceeded to the next phase of the test where we call the API method. In that second phase, we just verify that the state of the client was not changed because of the runtime failure. This tests handles runtime failures differently as it lets them bubble out whereas the previous test cases verifies that normal solr failures are handled internally and triggers a state change. 

This is why it is 2 different test cases instead of simply being another case for the previous test.